File: 1048 - Longest String Chain/solution.cpp
class Solution {
public:
    int longestStrChain(vector<string>& A) {
        int res = 0;
        sort(A.begin(), A.end(), [](auto& a, auto& b){return a.size()<b.size();});
        unordered_map<string, int> m;
        for(auto& s : A){
            for(int i = 0; i < s.size(); i++){
                string pre = s.substr(0, i)+s.substr(i+1);
                m[s] = max(m[s], m.count(pre)?m[pre]+1:1);
            }
            res = max(res, m[s]);
        }
        return res;
    }
};

File: 1048 - Longest String Chain/solution.py
class Solution:
    def longestStrChain(self, A: List[str]) -> int:
        res, m = 0, {}
        A.sort(key=len)
        for s in A:
            m[s] = 1
            for i in range(len(s)):
                pre = s[:i]+s[i+1:]
                if pre in m:
                    m[s] = max(m[s], m[pre]+1)
            res = max(res, m[s])
        return res